pkg/sql/alter_sequence.go

```diff
@@ -112,34 +114,56 @@ func (n *alterSequenceNode) startExec(params runParams) error {
 	// value of the sequence must be restored to the original MinValue or MaxValue transactionally.
 	// The code below handles the second case.
 
-	// The sequence is decreasing and the minvalue is being decreased.
-	if opts.Increment < 0 && desc.SequenceOpts.MinValue < oldMinValue {
-		sequenceVal, err := getSequenceValue()
-		if err != nil {
-			return err
-		}
+	sequenceVal, err := getSequenceValue()
+	if err != nil {
+		return err
+	}
 
-		// If the sequence exceeded the old MinValue, it must be changed to start at the old MinValue.
-		if sequenceVal < oldMinValue {
-			err := params.p.txn.Put(params.ctx, seqValueKey, oldMinValue)
+	// If the sequenceVal exceeded the old MinValue or MaxValue,
+	// we must set sequenceVal to the last valid one.
+
+	// I think we should set the current value to the newStartValue if there is
+	// a start param in alter_sequence_stmt, but in pg it needs RESTART.
+	// Now newStartValue has no effect on the current value before use RESTART.
+	// see: https://github.com/cockroachdb/cockroach/issues/74496
+	hasRestart := false
+	if hasRestart {
+		sequenceVal = opts.Start - opts.Increment
+	} else {
+		// 1. Values is just init, when the sequence is altered
+		// before it has ever been used.
+		if sequenceVal == oldStart-oldIncrement {
+			sequenceVal = oldStart - opts.Increment
+		} else {
+			// 2. The sequence is altered after it has ever been used.
+			currentVal, err := params.extendedEvalCtx.EvalContext.Sequence.GetLatestValueInSessionForSequenceByID(
```

Comment:
I'm now confused about how to get the current value of the sequence

Reply:
Since the sequenceVal obtained from seqValueKey is the last value generated in the cache, and we need the current sequence value, I added a currentSeqKey to store it.